sql/src/main/java/io/crate/planner/operators/Join.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.operators;
+
+import io.crate.analyze.OrderBy;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.expression.symbol.Field;
+import io.crate.expression.symbol.FieldsVisitor;
+import io.crate.expression.symbol.SelectSymbol;
+import io.crate.expression.symbol.Symbol;
+import io.crate.planner.node.dql.join.JoinType;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+abstract class Join extends TwoInputPlan {
+
+    final boolean hasOuterJoins;
+    AnalyzedRelation leftRelation;
+    AnalyzedRelation rightRelation;
+    protected Symbol joinCondition;
+
+    enum Side {
+        LEFT,
+        RIGHT
+    }
+
+    Join(LogicalPlan left,
+         LogicalPlan right,
+         List<Symbol> outputs,
+         boolean hasOuterJoins,
+         AnalyzedRelation leftRelation,
+         AnalyzedRelation rightRelation,
+         Symbol joinCondition) {
+        super(left, right, outputs);
+        this.hasOuterJoins = hasOuterJoins;
+        this.leftRelation = leftRelation;
+        this.rightRelation = rightRelation;
+        this.joinCondition = joinCondition;
+    }
+
+    protected abstract Side sidePreservingOrder();
+
+    protected abstract JoinType joinType();
+
+    Symbol joinCondition() {
+        return joinCondition;
+    }
+
+    @Override
+    public Map<LogicalPlan, SelectSymbol> dependencies() {
+        HashMap<LogicalPlan, SelectSymbol> deps = new HashMap<>();
+        deps.putAll(lhs.dependencies());
+        deps.putAll(rhs.dependencies());
+        return deps;
+    }
+
+    @Override
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown instanceof Order) {
```

Comment:
Again a bit of nitpicking against inheritance. I think we could extract the tryOptimize logic into a static functions within a `JoinOperators` and then avoid this Join base class. The operator class hierarchy is already at 4 levels - I would appreciate it if we could avoid making it bigger.

The `JoinOperators` class could also contain the helper functions that are now in the `MergePhase` which don't really belong there.